src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -1925,10 +1906,15 @@ MulticopterPositionControl::control_position(float dt)
 			thrust_sp(2) *= att_comp;
 		}
 
-		/* limit max thrust */
-		thrust_abs = thrust_sp.length(); /* recalculate because it might have changed */
+		/* Calculate desired total thrust amount in body z direction
+		 * Project the desired thrust force vector F onto the vehicles thrust axis -R_z
+		 * to compensate for excess thrust during attitude tracking errors */
+		matrix::Vector3f R_z(_R(0, 2), _R(1, 2), _R(2, 2));
+		matrix::Vector3f F(thrust_sp.data);
+		thrust_body_z = F.dot(-R_z); /* recalculate because it might have changed */
 
```

Comment:
A normal multicopter can only generate total thrust in body z direction.
I added some comments explaining the calculation I did.